nix/supervisord-cluster/clusterlib.py

```diff
@@ -1,167 +1,369 @@
-        self.CLI(["cardano-cli", "shelley", "transaction", "sign"] + sign_args )
-
-        # Submit TX
-        try:
-            self.CLI(["cardano-cli", "shelley", "transaction", "submit", "--testnet-magic", str(self.networkMagic), "--tx-file", "tx.signed"])
-        except CLIError as e:
-            print("Failed to submit transaction!")
-            print(f"txins: {txins} txouts: {txouts} signing keys: {signing_keys}")
-            raise e
-
-    def getAddress(self, payment=None, stake=None):
-        args = []
-        if not payment and stake:
-            raise CLIError("Must set payment, stake or both")
-        elif payment:
-            args.extend("--payment-verification-key-file", payment)
+        total_input_amount = functools.reduce(lambda x, y: x + y[2], txins, 0)
+        txouts_copy[-1] = (self.genesis_utxo_addr, (total_input_amount - fee - deposit_amount))
+        txins_combined = [f"{x[0]}#{x[1]}" for x in txins]
+        txouts_combined = [f"{x[0]}+{x[1]}" for x in txouts_copy]
+
+        txin_args = self.prepend_flag("--tx-in", txins_combined)
+        txout_args = self.prepend_flag("--tx-out", txouts_combined)
+        cert_args = self.prepend_flag("--certificate-file", certificates)
+
+        build_args = [
+            "cardano-cli",
+            "shelley",
+            "transaction",
+            "build-raw",
+            "--ttl",
+            str(ttl),
+            "--fee",
+            str(fee),
+            "--out-file",
+            str(out_file),
+            *txin_args,
+            *txout_args,
+            *cert_args,
+        ]
+
+        if proposal_file:
+            build_args.extend(["--update-proposal-file", proposal_file])
+
+        self.cli(build_args)
+
+    def sign_tx(self, tx_body_file="tx.body", out_file="tx.signed", signing_keys=None):
+        signing_keys = signing_keys or []
+        key_args = self.prepend_flag("--signing-key-file", signing_keys)
+        self.cli(
+            [
+                "cardano-cli",
+                "shelley",
+                "transaction",
+                "sign",
+                "--tx-body-file",
+                str(tx_body_file),
+                "--out-file",
+                str(out_file),
+                "--testnet-magic",
+                str(self.network_magic),
+                *key_args,
+            ]
+        )
+
+    def submit_tx(self, tx_file="tx.signed"):
+        self.cli(
+            [
+                "cardano-cli",
+                "shelley",
+                "transaction",
+                "submit",
+                "--testnet-magic",
+                str(self.network_magic),
+                "--tx-file",
+                str(tx_file),
+            ]
+        )
+
+    def get_address(self, payment=None, stake=None):
+        cli_args = []
+        if not (payment or stake):
+            raise CLIError("Must set payment, stake or both.")
```

Comment:
I think the condition should actually just be `not payment`.

The `shelley address build` command _must_ be provided with a `payment-verification-key-file`, while the `stake-verification-key-file` is optional.

Reply:
Also, since this method only utilizes the `shelley address build` command, it might be more accurate to call this method `get_payment_address`.

Unless there are plans to extend it to utilize the `shelley stake-address build` command in the future, @disassembler? e.g. in the case where _only_ a `stake-verification-key-file` is provided?

Reply:
Thanks for info! I changed it accordingly so that `stake` is now used alongside `payment` when specified.